pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/ScalarTransformFunctionWrapper.java

```diff
@@ -0,0 +1,301 @@
+      for (int k = 0; k < numNonLiteralArgs; k++) {
+        _args[_nonLiteralArgIndices.get(k)] = nonLiteralBlockValues[k][i];
+      }
+      _floatResult[i] = (Float) _functionInvoker.process(_args);
+    }
+    return _floatResult;
+  }
+
+  @SuppressWarnings("Duplicates")
+  @Override
+  public double[] transformToDoubleValuesSV(ProjectionBlock projectionBlock) {
+    if (_doubleResult == null) {
+      _doubleResult = new double[DocIdSetPlanNode.MAX_DOC_PER_CALL];
+    }
+    int length = projectionBlock.getNumDocs();
+    int numNonLiteralArgs = _nonLiteralArgIndices.size();
+    Object[][] nonLiteralBlockValues = new Object[numNonLiteralArgs][];
+
+    transformNonLiteralArgsToValues(projectionBlock, numNonLiteralArgs, nonLiteralBlockValues);
+
+    //now invoke the actual function
+    for (int i = 0; i < length; i++) {
+      for (int k = 0; k < numNonLiteralArgs; k++) {
+        _args[_nonLiteralArgIndices.get(k)] = nonLiteralBlockValues[k][i];
+      }
+      _doubleResult[i] = (Double) _functionInvoker.process(_args);
+    }
+    return _doubleResult;
+  }
+
+  @SuppressWarnings("Duplicates")
+  @Override
+  public String[] transformToStringValuesSV(ProjectionBlock projectionBlock) {
+    if (_stringResult == null) {
+      _stringResult = new String[DocIdSetPlanNode.MAX_DOC_PER_CALL];
+    }
+
+    int length = projectionBlock.getNumDocs();
+    int numNonLiteralArgs = _nonLiteralArgIndices.size();
+    Object[][] nonLiteralBlockValues = new Object[numNonLiteralArgs][];
+
+    transformNonLiteralArgsToValues(projectionBlock, numNonLiteralArgs, nonLiteralBlockValues);
+
+    //now invoke the actual function
+    for (int i = 0; i < length; i++) {
+      for (int k = 0; k < numNonLiteralArgs; k++) {
+        _args[_nonLiteralArgIndices.get(k)] = nonLiteralBlockValues[k][i];
+      }
+      _stringResult[i] = (String) _functionInvoker.process(_args);
+    }
+
+    return _stringResult;
+  }
+
+  private void transformNonLiteralArgsToValues(ProjectionBlock projectionBlock, int numNonLiteralArgs,
+      Object[][] nonLiteralBlockValues) {
+    for (int i = 0; i < numNonLiteralArgs; i++) {
+      TransformFunction transformFunc = _nonLiteralTransformFunction.get(i);
+      FieldSpec.DataType returnType = _nonLiteralArgType.get(i);
+      switch (returnType) {
+        case STRING:
+          nonLiteralBlockValues[i] = transformFunc.transformToStringValuesSV(projectionBlock);
+          break;
+        case INT:
+          int[] values = transformFunc.transformToIntValuesSV(projectionBlock);
+          nonLiteralBlockValues[i] = Arrays.stream(values).boxed().toArray(Integer[]::new);
+          break;
+        case DOUBLE:
+          double[] doubleValues = transformFunc.transformToDoubleValuesSV(projectionBlock);
+          nonLiteralBlockValues[i] = Arrays.stream(doubleValues).boxed().toArray(Double[]::new);
+          break;
+        case FLOAT:
+          float[] floatValues = transformFunc.transformToFloatValuesSV(projectionBlock);
```

Comment:
Why using different way to box the values?
I would suggest changing all types to be the same as FLOAT. From the past experience, seems the streaming fashion has worse performance.

Reply:
Another way is using `org.apache.commons.lang3.ArrayUtils.toObject()` which is clean and neat